components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorage.kt

```diff
@@ -27,26 +33,34 @@ import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
 class SyncedTabsStorage(
     private val accountManager: FxaAccountManager,
     private val store: BrowserStore,
-    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage()
+    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage(),
+    private val debounceMillis: Long = 1000L,
 ) : SyncedTabsProvider {
     private var scope: CoroutineScope? = null
 
     /**
      * Start listening to browser store changes.
      */
+    @OptIn(FlowPreview::class)
     fun start() {
         scope = store.flowScoped { flow ->
-            flow.ifChanged { it.tabs }.map { state ->
+            flow.ifChanged { it.toSyncComponents() }.map { state ->
                 // TO-DO: https://github.com/mozilla-mobile/android-components/issues/5179
                 val iconUrl = null
-                state.tabs.filter { !it.content.private }.map { tab ->
+                state.tabs.filter { !it.content.private && !it.content.loading }.map { tab ->
                     // TO-DO: https://github.com/mozilla-mobile/android-components/issues/1340
                     val history = listOf(TabEntry(tab.content.title, tab.content.url, iconUrl))
                     Tab(history, 0, tab.lastAccess)
                 }
-            }.collect { tabs ->
-                tabsStorage.store(tabs)
             }
+                .debounce(debounceMillis)
```

Comment:
(not sure if I missed this line the first pass)

With the account manager's debouncing mechanism that already exists, I'm not sure there is more added value here any more. Saving on disk writes means potentially losing state that we could be syncing the next time the device is able to.

Reply:
The account manager's debouncing mechanism seems to operate in the opposite manner to the `Flow.debounce` operator. That is:

- flow debouncing emits the latest value based on the timeout
- sync manager will ignore additional sync requests if they are within a debounce period

I think we want the behavior of the of flow debouncing, regardless of whether we want to use this specific implementation. Tab changes can happen rapidly due to different actions, so we want to wait for a "final" state before initiating a sync. 

If we aren't concerned about disk write frequency, I can rework the debounce mechanism to operate specifically on the sync requests

Reply:
> I think we want the behavior of the of flow debouncing, regardless of whether we want to use this specific implementation. Tab changes can happen rapidly due to different actions, so we want to wait for a "final" state before initiating a sync.

Wouldn't the sync manager's debounce support this behaviour too? We get rapid tab changes, the sync manager ignores most requests until the debounce times out and the next sync is allowed. Is that the right understanding? 

>If we aren't concerned about disk write frequency, I can rework the debounce mechanism to operate specifically on the sync requests

I think we only need to remove the flow `debounce`. 🤔

Reply:
One thing to consider is that the sync manager debounce timer is also 5 minutes, and isn't currently configurable. My understanding of the sync manager debounce could lead to this situation:

- storage observes initial state, calls `syncNow(debounce = true)`
- unloaded tabs load
- selected tab is changed (any number of times within 5 minutes)

We would debounce every change that happened _after_ the initial state load until the 5 minute window had passed. I think we want the opposite: to ignore every change during a window until the last one, so we sync the most recent data.

Reply:
Thanks for the explanation and doing the homework to study the differences between the two. I think your analysis is correct and we're better off with both these debouncers.